src/main/java/ch/uzh/ifi/hase/soprafs24/service/TaskService.java

```diff
@@ -80,71 +71,106 @@ public Task createTask(Task newTask, long userId) {
         }
-        Application newApplication = new Application();
-        newApplication.setTask(selectedTask);
-        newApplication.setUser(candidate);
-        applicationsRepository.saveAndFlush(newApplication);
-    }
-
-    public void selectCandidate(TaskPutDTO taskPutDTO, String token){
-        // QUESTION DANA. AFTER ANSWER DELETE COMMENTS IN DTOMAPPER. necessary because the user in the task entity is saved as an entity and is not mappable with
-        //taskputdto since there the userId is a long
-        User helper = this.userRepository.findUserById(taskPutDTO.getHelperId());
-        User taskCreator = userRepository.findUserById(taskPutDTO.getUserId());
-        Task task = taskRepository.findById(taskPutDTO.getTaskId());
-        Application application = this.applicationsRepository.findByUserAndTask(helper, task);
-
-        //Check the taskCreator is performing the selection action
-        if (!taskCreator.getToken().equals(token)){
+
+        // Add the candidate to the task
+        selectedTask.addCandidate(candidate);
+
+        // Save the updated task
+        taskRepository.save(selectedTask);
+    }
+
+
+    @Transactional
+    public void selectCandidate(TaskPutDTO taskPutDTO, String token) {
+        User helper = userService.getUserById(taskPutDTO.getHelperId());
+        User taskCreator = userService.getUserById(taskPutDTO.getUserId());
+        Task task = taskRepository.findById(taskPutDTO.getTaskId()).orElseThrow(() ->
+                new ResponseStatusException(HttpStatus.NOT_FOUND, "The task was not found."));
+
+        // Check the taskCreator is performing the selection action
+        if (!taskCreator.getToken().equals(token)) {
             throw new ResponseStatusException(HttpStatus.FORBIDDEN, "Only the creator of a task can choose the helper");
         }
-        //Check the task was retrieved correctly
-        if (task == null){
-            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "The task was not found.");
-        }
-        //Check the application actually exists
-        if (application == null){
+
+        // Check if the helper is a valid candidate for the task
+        if (!task.hasCandidate(helper)) {
             throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "This helper has not applied for the job.");
         }
 
+        // Set the helper and update the task status
         task.setHelper(helper);
         task.setStatus(TaskStatus.IN_PROGRESS);
-        deleteApplicationsByTask(task, helper);
+
+        // Clear other applications
+        clearOtherCandidates(task, helper);
+
+        taskRepository.save(task); // Save the task with the new helper and status
     }
 
-    public void deleteTaskWithId(long taskId, String token) {
-        Task taskToBeDeleted = this.taskRepository.findById(taskId);
-        if (taskToBeDeleted == null) {
-            throw new NoSuchElementException("Task not found with id: " + taskId);
```

Comment:
the clearOtherCandidates replaces the deleteApplicationsByTask. The name makes it clear that other candidates, except the one passed as argument, are removed from the task. It does not use the applicationRepository anymore, removes the task from the candidates' applications and the candidates from the task's applications.